File: src-lang/js/ui/figma.clj
(ns js.ui.figma
  "React renderer for portable xt.ui nodes using @xtalk/figma-ui."
  (:require [hara.lang :as l]))

(l/script :js
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.common-data :as xtd]
             [xt.ui.core :as ui]
             [xt.ui.catalog :as catalog]
             [xt.ui.model :as ui-model]
             [js.react :as r]
             [js.lib.figma :as figma]]})

(defn.js normalize-props
  "maps portable property/event names onto React conventions"
  [props]
  (var out (xt/x:obj-assign {} (or props {})))
  (when (xt/x:has-key? out "class")
    (xt/x:set-key out "className" (xt/x:get-key out "class"))
    (xt/x:del-key out "class"))
  (when (xt/x:has-key? out "aria_label")
    (xt/x:set-key out "aria-label" (xt/x:get-key out "aria_label"))
    (xt/x:del-key out "aria_label"))
  (when (xt/x:has-key? out "read_only")
    (xt/x:set-key out "readOnly" (xt/x:get-key out "read_only"))
    (xt/x:del-key out "read_only"))
  (when (xt/x:has-key? out "for")
    (xt/x:set-key out "htmlFor" (xt/x:get-key out "for"))
    (xt/x:del-key out "for"))
  (when (xt/x:has-key? out "tone")
    (xt/x:set-key out "variant"
                  (:? (== "error" (xt/x:get-key out "tone"))
                      "destructive"
                      "default"))
    (xt/x:del-key out "tone"))
  (when (xt/x:has-key? out "on_press")
    (xt/x:set-key out "onClick" (xt/x:get-key out "on_press"))
    (xt/x:del-key out "on_press"))
  (when (xt/x:has-key? out "on_submit")
    (xt/x:set-key out "onSubmit" (xt/x:get-key out "on_submit"))
    (xt/x:del-key out "on_submit"))
  (when (xt/x:has-key? out "on_change")
    (var callback (xt/x:get-key out "on_change"))
    (xt/x:set-key
     out "onChange"
     (fn [event]
       (return (callback (xtd/get-in event ["target" "value"])))))
    (xt/x:del-key out "on_change"))
  (xt/x:del-key out "pending")
  (xt/x:del-key out "hidden")
  (return out))

(defn.js web-registry
  "creates the default React registry; app layers can override any renderer"
  []
  (var platform (ui/registry-create "xt.ui/react-figma"))
  (ui/registry-register-renderer platform "ui/fragment" r/Fragment)
  (ui/registry-register-renderer platform "ui/row" "div")
  (ui/registry-register-renderer platform "ui/column" "div")
  (ui/registry-register-renderer platform "ui/text" "span")
  (ui/registry-register-renderer platform "ui/icon" "span")
  (ui/registry-register-renderer platform "ui/image" "img")
  (ui/registry-register-renderer platform "ui/card" figma/Card)
  (ui/registry-register-renderer platform "ui/card-header" figma/CardHeader)
  (ui/registry-register-renderer platform "ui/card-content" figma/CardContent)
  (ui/registry-register-renderer platform "ui/title" figma/CardTitle)
  (ui/registry-register-renderer platform "ui/description" figma/CardDescription)
  (ui/registry-register-renderer platform "ui/label" figma/Label)
  (ui/registry-register-renderer platform "ui/input" figma/Input)
  (ui/registry-register-renderer platform "ui/textarea" figma/Textarea)
  (ui/registry-register-renderer platform "ui/button" figma/Button)
  (ui/registry-register-renderer platform "ui/alert" figma/Alert)
  (ui/registry-register-renderer platform "ui/spinner" figma/Skeleton)
  (return (ui/registry-compose [(catalog/registry) platform])))

(defn.js render-node
  "renders a portable node recursively through the runtime registry"
  [runtime value]
  (when (xt/x:nil? value)
    (return nil))
  (when (or (xt/x:is-string? value) (xt/x:is-number? value))
    (return value))
  (when (xt/x:is-array? value)
    (return (xtd/arr-map value (fn [child]
                                (return (-/render-node runtime child))))))
  (var component-id (xt/x:get-key value "component"))
  (when (== component-id "ui/slot")
    (return (-/render-node runtime (ui/resolve-slot runtime value))))
  (var props (xt/x:get-key value "props"))
  (when (== true (xt/x:get-key props "hidden"))
    (return nil))
  (var renderer (ui/registry-renderer (xt/x:get-key runtime "registry") component-id))
  (when (not renderer)
    (xt/x:err (xt/x:cat "ERR - missing React UI renderer - " component-id)))
  (var children (-/render-node runtime (xt/x:get-key value "children")))
  (var rprops (-/normalize-props props))
  (when (== component-id "ui/row")
    (xt/x:set-key rprops "className"
                  (xt/x:cat "flex flex-row " (or (xt/x:get-key rprops "className") ""))))
  (when (== component-id "ui/column")
    (xt/x:set-key rprops "className"
                  (xt/x:cat "flex flex-col " (or (xt/x:get-key rprops "className") ""))))
  (when (or (== component-id "ui/text")
            (== component-id "ui/title")
            (== component-id "ui/description")
            (== component-id "ui/label"))
    (:= children [(xt/x:get-key props "value")])
    (xt/x:del-key rprops "value"))
  (return (r/createElement renderer rprops children)))

(defn.js use-model-store
  "binds React rendering to the same keyed model events used by proxy/local nodes"
  [store subscription-id]
  (r/useSyncExternalStore
   (fn [notify]
     (ui-model/subscribe! store subscription-id notify)
     (return (fn [] (ui-model/unsubscribe! store subscription-id))))
   (fn [] (return (ui-model/store-version store)))
   (fn [] (return (ui-model/store-version store))))
  (return store))

(defn.js PortableView
  [#{runtime view-fn subscription-id}]
  (-/use-model-store (xt/x:get-key runtime "store")
                     (or subscription-id "xt.ui/react-view"))
  (return (-/render-node runtime (view-fn runtime))))

File: packages-gen/js/ui/ui/figma.js
const React = require("react")

const FigmaUi = require("@xtalk/figma-ui")

const ui = require("@xtalk/ui/core.js")

const xtd = require("@xtalk/lang/common-data.js")

const ui_page = require("@xtalk/ui/page.js")

const catalog = require("@xtalk/ui/catalog.js")

const ui_model = require("@xtalk/ui/model.js")

function normalize_props(props){
  let out = Object.assign({},props || {});
  if(null != out["class"]){
    out["className"] = out["class"];
    delete(out["class"]);
  }
  if(null != out["aria_label"]){
    out["aria-label"] = out["aria_label"];
    delete(out["aria_label"]);
  }
  if(null != out["read_only"]){
    out["readOnly"] = out["read_only"];
    delete(out["read_only"]);
  }
  if(null != out["for"]){
    out["htmlFor"] = out["for"];
    delete(out["for"]);
  }
  if(null != out["tone"]){
    out["variant"] = (("error" == out["tone"]) ? "destructive" : "default");
    delete(out["tone"]);
  }
  if(null != out["on_press"]){
    out["onClick"] = out["on_press"];
    delete(out["on_press"]);
  }
  if(null != out["on_submit"]){
    out["onSubmit"] = out["on_submit"];
    delete(out["on_submit"]);
  }
  if(null != out["on_change"]){
    let callback = out["on_change"];
    out["onChange"] = (function (event){
      return callback(xtd.get_in(event,["target","value"]));
    });
    delete(out["on_change"]);
  }
  delete(out["pending"]);
  delete(out["hidden"]);
  return out;
}

function web_registry(){
  let platform = ui.registry_create("xt.ui/react-figma");
  ui.registry_register_renderer(platform,"ui/fragment",React.Fragment);
  ui.registry_register_renderer(platform,"ui/row","div");
  ui.registry_register_renderer(platform,"ui/column","div");
  ui.registry_register_renderer(platform,"ui/text","span");
  ui.registry_register_renderer(platform,"ui/icon","span");
  ui.registry_register_renderer(platform,"ui/image","img");
  ui.registry_register_renderer(platform,"ui/card",FigmaUi.Card);
  ui.registry_register_renderer(platform,"ui/card-header",FigmaUi.CardHeader);
  ui.registry_register_renderer(platform,"ui/card-content",FigmaUi.CardContent);
  ui.registry_register_renderer(platform,"ui/title",FigmaUi.CardTitle);
  ui.registry_register_renderer(platform,"ui/description",FigmaUi.CardDescription);
  ui.registry_register_renderer(platform,"ui/label",FigmaUi.Label);
  ui.registry_register_renderer(platform,"ui/input",FigmaUi.Input);
  ui.registry_register_renderer(platform,"ui/textarea",FigmaUi.Textarea);
  ui.registry_register_renderer(platform,"ui/button",FigmaUi.Button);
  ui.registry_register_renderer(platform,"ui/alert",FigmaUi.Alert);
  ui.registry_register_renderer(platform,"ui/spinner",FigmaUi.Skeleton);
  return ui.registry_compose([catalog.registry(),platform]);
}

function render_node(runtime,value){
  if(null == value){
    return null;
  }
  if(("string" == (typeof value)) || ("number" == (typeof value))){
    return value;
  }
  if(Array.isArray(value)){
    return xtd.arr_map(value,function (child){
      return render_node(runtime,child);
    });
  }
  let component_id = value["component"];
  if(component_id == "ui/slot"){
    return render_node(runtime,ui.resolve_slot(runtime,value));
  }
  let props = value["props"];
  if(true == props["hidden"]){
    return null;
  }
  let renderer = ui.registry_renderer(runtime["registry"],component_id);
  if(!renderer){
    throw "ERR - missing React UI renderer - " + component_id;
  }
  let children = render_node(runtime,value["children"]);
  let rprops = normalize_props(props);
  if(component_id == "ui/row"){
    rprops["className"] = ("flex flex-row " + (rprops["className"] || ""));
  }
  if(component_id == "ui/column"){
    rprops["className"] = ("flex flex-col " + (rprops["className"] || ""));
  }
  if((component_id == "ui/text") || (component_id == "ui/title") || (component_id == "ui/description") || (component_id == "ui/label")){
    children = [props["value"]];
    delete(rprops["value"]);
  }
  return React.createElement(renderer,rprops,children);
}

function use_model_store(store,subscription_id){
  React.useSyncExternalStore(function (notify){
    ui_model.subscribef(store,subscription_id,notify);
    return function (){
      ui_model.unsubscribef(store,subscription_id);
    };
  },function (){
    return ui_model.store_version(store);
  },function (){
    return ui_model.store_version(store);
  });
  return store;
}

function PortableView({runtime,subscription_id,view_fn}){
  use_model_store(runtime["store"],subscription_id || "xt.ui/react-view");
  return render_node(runtime,view_fn(runtime));
}

function use_page_controller(controller,subscription_id){
  React.useSyncExternalStore(function (notify){
    ui_page.subscribef(controller,subscription_id,function (_state,_revision){
      notify();
    });
    return function (){
      ui_page.unsubscribef(controller,subscription_id);
    };
  },function (){
    return ui_page.revision(controller);
  },function (){
    return ui_page.revision(controller);
  });
  return ui_page.snapshot(controller);
}

function PortableControllerView({actions,controller,runtime,subscription_id,view_fn}){
  let state = use_page_controller(controller,subscription_id || "xt.ui/react-controller-view");
  return render_node(runtime,view_fn(state,actions || {}));
}

module.exports = {
  ["normalize_props"]:normalize_props,
  ["web_registry"]:web_registry,
  ["render_node"]:render_node,
  ["use_model_store"]:use_model_store,
  ["PortableView"]:PortableView,
  ["use_page_controller"]:use_page_controller,
  ["PortableControllerView"]:PortableControllerView
}